File: C/turing.py
#!/usr/bin/python

# #################################################################
__license__= '''This module (version Mar/2015),
comes with ABSOLUTELY NO WARRANTY.  This is free software, and you are welcome to 
redistribute it under conditions of the GNU General Public License.'''
__author__ = "NelsonManohar"
__date__   = "Mar2015"
__doc__    = '''a pure python turing machine tape reader handling extended
finite state machine automata for the printable ascii character set, 
assumptions given: tape is not infinite nor very long
assumption: tape initial indexing starts at 0'''
print '-' * 80
print __doc__
print '-' * 80
print __license__
print '-' * 80
# #################################################################


# #################################################################
from collections import defaultdict
import time
import argparse
import sys
import string
# #################################################################


# #################################################################
class TAPE_READER_CLASS(object):
    # ###########################################################
    def __init__(self, tapefile, maxiter=0, debug=1):
        self.DEBUG = debug
        self.TAPEFILE = tapefile
        self.MAXITER = maxiter
        self.NITER = 0
        self.TAPE = None
        self.ACTION_TABLE = defaultdict(list)
        self.TAPE_STATE = dict(zip(('contents','blank','position','ps_state','hs_state', 'direction'),
                                    (None,      '-',    None,      None,      None,      None)))
        self.DONE = False
        self.MATCHING_STATE = []
        #self.load_tape_header()
        #self.load_action_table()
        #self.print_action_table()
        #self.main()
        return
    # ###########################################################

    # ###########################################################
    def INFO(self, msg, val):
        print '-' * 80
        print "INFO   : [%s] [%s]" % (msg, val)
        return val
    # ###########################################################

    # ###########################################################
    def ERROR(self, msg, val):
        print "ERROR  : [%s] [%s]" % (msg, val)
        self.done(-1)
        return val
    # ###########################################################

    # ###########################################################
    def WARNING(self, msg, val ):
        print "WARNING: [%s] [%s]" % (msg, val)
        return val
    # ###########################################################

    # ###########################################################
    def print_action_table(self):
        print '-' * 80
        print "[%s] [%s] THEN write char [%s], update dir to [%s];\t set next state -> [%s]" % ( "PS", "C", "W", "D", "NS" ) 
        print '-' * 80
        akeys = sorted(self.ACTION_TABLE.keys())
        for ps in akeys:
            for actions in self.ACTION_TABLE[ps]:
                print "[%s] [%s]\t THEN write char [%s], update dir to [%s];\t set next state -> [%s]" % tuple(actions)
        print '-' * 80
    # ###########################################################

    # ###########################################################
    def load_action_table(self):
        '''
        loads an tape reader automata action table, where each row of an action table 
        automata specification is expected to match this exacting format
                <State index> <If character> <Write> <Direction> <New state index>
        '''
        self.ACTION_TABLE  = defaultdict(list)
        try:
            with open(self.TAPEFILE) as tape:
                tape.readline()
                tape.readline()
                tape.readline()
                tape.readline()
                done = False
                while not done:
                    try:
                        line = tape.readline().strip()
                        if not line: break
                        (ps, ifchar, writechar, direction, ns) = line.split(' ')
                        ps, direction, ns = int(ps), int(direction), int(ns)
                    except:
                        done = True
                        self.ERROR("ACTION TABLE LINE IS WRONG", line)
                    self.ACTION_TABLE[ps].append([ps, ifchar, writechar, direction, ns])
    
                if not self.ACTION_TABLE:
                    self.ERROR("ACTION TABLE SPECIFICATION MISSING", self.TAPEFILE)
        except:
            self.ERROR("ACTION TABLE SPECIFICATION IS WRONG", self.TAPEFILE)
        return self.ACTION_TABLE
    # ###########################################################
    
    # ###########################################################
    def load_tape_header(self):
        ''' Expects this format:
        <Starting contents of tape>\n
        <Index into above string determining start position of head>\n
        <Start state index (integer)>\n
        <Halting state index (integer)>\n
        <Action table>
        '''
        try:
            with open(self.TAPEFILE) as tape:
                chars = tape.readline().strip()
                pos = tape.readline().strip()
                ps = tape.readline().strip()
                hs = tape.readline().strip()
                if self.DEBUG:
                    print "%s\n, %s %s %s " % (chars, pos, ps, hs)
        except:
            self.ERROR("TAPE HEADER FORMAT IS WRONG", self.TAPEFILE)

        self.TAPE_STATE['contents'] = chars
        self.TAPE_STATE['tape_out'] = []
        try:
            self.TAPE_STATE['position'] = int(pos)
            self.TAPE_STATE['ps_state'] = int(ps)
            self.TAPE_STATE['hs_state'] = int(hs)
        except:
            self.ERROR("TAPE HEADER CONTAINS INVALID STATE/POSITION, INTEGER EXPECTED", (pos,ps,hs))

        if not all(x in string.printable for x in chars):
            self.ERROR("TAPE CONTAINS NON PRINTABLE CHARACTERS", chars)

        return self.TAPE_STATE
    # ###########################################################

    # ###########################################################
    def readchar(self):
        '''Assumes tape head is already positioned and returns proper character'''
        ifchar = None
        self.NITER += 1
        try:
            pos = self.TAPE_STATE['position']
            if pos < 0 or pos > len(self.TAPE_STATE['contents']):
                self.ERROR("INVALID TAPE POSITION, REACHED OUTSIDE TAPE BOUNDARIES", pos)
            ifchar = self.TAPE_STATE['contents'][int(pos)]
            print "[ITERATION: %s] AT POS[%s], READ [%s]" % (self.NITER, pos, ifchar)
        except:
            self.ERROR("INVALID TAPE POSITION", pos)
        return ifchar
    # ###########################################################
    
    # ###########################################################
    def next_state(self, char):
        ps = self.TAPE_STATE['ps_state']
        states = self.ACTION_TABLE[ps]

        matching_state = ()
        for state in states:
            (ps, ifchar, writechar, direction, ns) = state
            if char == ifchar:
                matching_state = state[:]
                break
            if ifchar == "*":
                writechar = ""      # if matching state is *, then write NOTHING as specified
                matching_state = (ps, ifchar, writechar, direction, ns)
                break

        if not matching_state:
            self.ERROR("ACTION TABLE LACKS RULE FOR", "PS=%s IFCHAR=[%s])" % (ps, char))

        self.MATCHING_STATE = matching_state[:]
        return self.MATCHING_STATE
    # ###########################################################
    
    # ###########################################################
    def update_state(self, char, action_state):
        (ps, ifchar, writechar, direction, ns) = action_state
        if self.DEBUG:
            print "(%s:(%s->%s))->(%s,%s)" % (ps, char, ifchar, writechar, ns)
        self.TAPE_STATE['direction'] = int(direction)
        self.TAPE_STATE['position'] = int(self.TAPE_STATE['position']) + int(direction)
        self.TAPE_STATE['ps_state'] = ns
        self.TAPE_STATE['tape_out'].append((char,writechar))
        return
    # ###########################################################

    # ###########################################################
    def print_mapping(self):
        for i, pair in enumerate(self.TAPE_STATE['tape_out']):
            print "%6s: %s --> %s" % (i, pair[0], pair[1])
        return self.TAPE_STATE['tape_out']
    # ###########################################################

    # ###########################################################
    def halting_tests(self):
        if self.TAPE_STATE['ps_state'] not in self.ACTION_TABLE.keys():
            self.INFO("TAPE READER FOUND AT UNDEFINED ACTION STATE\n", self.TAPE_STATE)
            if self.TAPE_STATE['ps_state'] == self.TAPE_STATE['hs_state']:
                self.INFO("TAPE READER REACHED HALTING STATE\n", self.TAPE_STATE)
            self.DONE = True
            self.done(-1)
        if self.TAPE_STATE['ps_state'] == self.TAPE_STATE['hs_state']:
            self.INFO("TAPE READER REACHED HALTING STATE\n", self.TAPE_STATE)
            self.DONE = True
            self.done(0)
        if self.MAXITER:
            if self.NITER > self.MAXITER:
                self.INFO("MAXIMUM NUMBER OF ITERATIONS REACHED BEFORE HALTING STATE\n", self.TAPE_STATE)
                self.DONE = True
                self.done(1)
        return
    # ###########################################################

    # ###########################################################
    def done(self, how):
        print '-' * 80
        tapeInput = "".join([x for x in self.TAPE_STATE['contents']])
        tapeRead = "".join([x[0] for x in self.TAPE_STATE['tape_out']])
        tapeOutput = "".join([x[1] for x in self.TAPE_STATE['tape_out']])
        print "TAPE CONTENTS: ", tapeInput
        print "TAPE TRAVERSAL:", tapeRead
        print "OUTPUT STREAM: ", tapeOutput
        print '-' * 80
        if how != 0:
            sys.exit(how)
        return (tapeInput, tapeRead, tapeOutput)
    # ###########################################################

    # ###########################################################
    def main(self): 
        while not self.DONE:
            self.halting_tests()
            if not self.DONE:
                char = self.readchar()
                self.MATCHING_STATE = self.next_state(char)
                self.update_state(char, self.MATCHING_STATE)
                self.print_mapping()
        return self.TAPE_STATE


    # ###########################################################
# #################################################################

    
# #################################################################
if __name__ == "__main__":
    parser = argparse.ArgumentParser(prog='turing.py',
             description='This program operates a virtual tape reader wrt an extended finite state machine automata specification.')
    parser.add_argument("--tapefile",
             default='tapefile.txt', help='filepath to file containing the lines to be sampled.')
    parser.add_argument("--maxiter", type=float,
             default=0, help='if non zero, a max. number of traversals over the tape, for handling infinite loops.')
    args = parser.parse_args()
    parser.print_help()

    print >>sys.stderr, '-' * 80
    print >>sys.stderr, "tape file to scan:", args.tapefile
    print >>sys.stderr, '-' * 80

    tapeReader = TAPE_READER_CLASS(args.tapefile, maxiter=args.maxiter)
    tapeReader.load_tape_header()
    tapeReader.load_action_table()
    tapeReader.print_action_table()
    tapeReader.main()

File: C/turing.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <ctype.h>
#include <errno.h>

//#define ISPRINT(x) isprint(x)
#define ISPRINT(x) x >= 32 && x <= 127 ? 1 : 0
#define ISLETTER(x) isletter(x)
#define SLEN(x) strlen(x)
#define AND &&
#define OR &&
#define NOT !

/* constants */
const int NIL=0x0;              /* inits/defines nil pointer */
const int NOT_FOUND = -123456;
const int WILDCARD = 42;
const int NOTPRINT = 32;
const int MAXOUT = 100000;

/* arrays */
#define MSG(s,v) printf("[%12s]:\t %s\n", s, v);
#define HEADER() printf("%s", "------------------------------------------------------------\n");


/* ********************************************************************************************** */
/* prototypes */
/* ********************************************************************************************** */
extern char *filereader(char *);
extern char *readline(FILE *, int);
extern char readchar(void);
extern void load_tapefile(char *);
extern void print_action_table(void);
extern int nextstate(char);
extern void update_state(char, int);
extern int halting_tests(void);
extern void print_tape_state(void);
extern void msg(char *, char *, int);
/* ********************************************************************************************** */


/* ********************************************************************************************** */
// THE GLOBAL STATE OF THE TAPE
/* ********************************************************************************************** */
typedef struct TAPE_STATE_S {
    char *contents;
    char *traversal; 
    char *output; 
    int position;
    int pstate;
    int hstate;
    int direction;
} TAPE_STATE_T;

TAPE_STATE_T TAPE_STATE;
/* ********************************************************************************************** */


/* ********************************************************************************************** */
// THE CONTAINER DATA FOR EACH INDIVIDUAL ACTION TABLE ROW
/* ********************************************************************************************** */
struct STATE {
    int pstate;
    char ifchar;
    char writechar;
    int direction;
    int nstate;
    struct STATE *next;
} STATE[100];

static int NUM_STATES;
static int NUM_ITERS = 0;
static int OUTPUT_IDX = 0;
/* ********************************************************************************************** */


/* ********************************************************************************************** */
void
msg( char *type, char *msg, int code ) {
    printf( "[%12s] %d : %s\n", type, code, msg );
    if (type == "ERROR")
        exit(code);
    return;
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
/*derived from http://stackoverflow.com/questions/174531/easiest-way-to-get-files-contents-in-c */
/* ********************************************************************************************** */
void
load_tapefile(char *filename) {
    /* char *contents;
     * char *output; 
     * int position;
     * int pstate;
     * int hstate;
     * int direction;
     */

    FILE    *fp;
    int      length;
    char    *buffer;
    int      i, j;

    fp = fopen( filename, "r");
    if (fp == NULL) {
        fclose(fp);
        msg("ERROR", "CAN NOT OPEN SPECIFIED TAPE FILE", -1);
    }

    fseek (fp, 0, SEEK_END);
    length = ftell (fp);
    fseek (fp, 0, SEEK_SET);

    buffer = readline(fp, length);
    TAPE_STATE.contents = (char *) buffer;
    for (j=0; j<strlen(buffer); j++) {
        if (!ISPRINT((int) buffer[j])) {
            fclose(fp);
            msg("ERROR", "NON PRINTABLE ASCII CHARACTERS ON TAPE", -2);
        }
    }
    TAPE_STATE.output = (char *) malloc(MAXOUT);
    TAPE_STATE.traversal = (char *) malloc(MAXOUT);

    buffer = readline(fp, 80);
    sscanf(buffer, "%d", &TAPE_STATE.position);

    buffer = readline(fp, 80);
    sscanf(buffer, "%d", &TAPE_STATE.pstate);

    buffer = readline(fp, 80);
    sscanf(buffer, "%d", &TAPE_STATE.hstate);

    buffer = readline(fp, 80);
    i = 0;
    while ( buffer ) {
        if (!buffer) break;
        sscanf(buffer, "%d %c %c %d %d", &STATE[i].pstate, &STATE[i].ifchar, &STATE[i].writechar, &STATE[i].direction, &STATE[i].nstate);
        i++;
        buffer = readline(fp, 80);
    }

    NUM_STATES = i;

    fclose (fp);
    return;
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
void
print_action_table() {
    int i;
    HEADER();
    printf("ACTION TABLE\n");
    HEADER();
    printf("%s \t %s \t %s \t %s \t %s\n", "PS", "IFC", "WRTC", "DIR", "NS" );
    for (i=0; i<NUM_STATES; i++) {
        printf("%d \t %c \t %c \t %d \t %d\n", STATE[i].pstate, STATE[i].ifchar, STATE[i].writechar, STATE[i].direction, STATE[i].nstate);
    }
    HEADER();
    return;
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
char *
readline(FILE *fp, int length) {
    char *buffer;
    char *ptr;

    buffer = malloc (length);
    if (buffer) {
        ptr = fgets( buffer, length, fp );
    } 
    if (!buffer) {
        fclose(fp);
        msg("ERROR", "CAN NOT ALLOCATE MEMORY BUFFER", -3);
    }
    if (!ptr) {
        MSG("INFO", "READ FILE STREAM");
        return(NULL);
    }
    return(buffer);
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
char
readchar() {
    int pos;
    char tapechar;

    pos = TAPE_STATE.position;
    if (pos < 0 ) {
        msg( "ERROR", "TAPE HEAD OUTSIDE BOUNDARY", -4);
    }

    if (pos >= strlen(TAPE_STATE.contents)) {
        msg( "ERROR", "TAPE HEAD OUTSIDE BOUNDARY", -5);
    }

    tapechar = TAPE_STATE.contents[pos];
    printf( "ITER %d, PRESENT STATE [%d] AT POSITION = %d; READ CHAR %c\n", NUM_ITERS, TAPE_STATE.pstate, pos, tapechar);

    return tapechar;
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
int
nextstate(char c) {
    int ps;
    int idx;
    int found;

    ps = TAPE_STATE.pstate;

    found = NOT_FOUND;
    for (idx=0; idx<NUM_STATES; idx++) {
        if ( STATE[idx].pstate == ps )  {
            if (STATE[idx].ifchar == c) {
                found = idx;
                break;
            }
            if (STATE[idx].ifchar == WILDCARD) {
                found = idx;
                break;
            }
        }
    }

    if (found == NOT_FOUND ) {
        msg("ERROR", "REACHED INVALID STATE", -6);
    }

    printf("MATCHING RULE: PS=%d \t IFC=%c \t WRT=%c \t DIR=%d \t NS=%d\n", STATE[idx].pstate, STATE[idx].ifchar, STATE[idx].writechar, STATE[idx].direction, STATE[idx].nstate);
    return (found);
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
void
update_state(char c, int idx) {
    int ps, direction, nstate;
    char ifchar, writechar;

    ps = STATE[idx].pstate;
    ifchar = STATE[idx].ifchar;
    writechar = STATE[idx].writechar;
    direction = STATE[idx].direction;
    nstate = STATE[idx].nstate;

    TAPE_STATE.position = TAPE_STATE.position + direction;
    TAPE_STATE.direction = direction;
    TAPE_STATE.pstate = nstate;

    if (writechar == WILDCARD) writechar = NOTPRINT;

    TAPE_STATE.traversal[NUM_ITERS]= (char) c;
    TAPE_STATE.output[OUTPUT_IDX]  = (char) writechar;

    printf( "UPDATE: NEXT STATE = %d, POSITION = %d, DIRECTION = %d\n", TAPE_STATE.pstate, TAPE_STATE.position, TAPE_STATE.direction );
    if (writechar != WILDCARD) {
        printf( "GIVEN IFCHAR [%c], WROTE CHAR [%c]\n", ifchar, writechar );
    } else {
        printf( "GIVEN IFCHAR [%c], WROTE CHAR []\n", ifchar );
    }

    if (writechar != NOTPRINT) OUTPUT_IDX++;
    NUM_ITERS++;

    if (NUM_ITERS>=MAXOUT) {
        msg("ERROR", "CAN NOT OPEN SPECIFIED TAPE FILE", -29);
    }

    return;
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
int
halting_tests() {
    if (TAPE_STATE.pstate == TAPE_STATE.hstate) {
        MSG( "INFO", "HALTING STATE BEEN REACHED." );
        return (0);
    }
    return(1);
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
void 
print_tape_state() {
    printf( "TAPE CONTENTS : %s",   TAPE_STATE.contents);
    printf( "TAPE TRAVERSAL: %s\n", TAPE_STATE.traversal);
    printf( "OUTPUT STREAM : %s\n", TAPE_STATE.output);
    return;
}
/* ********************************************************************************************** */


/* ********************************************************************************************** */
void main( int argc, char **argv) {
    char *tapefile;
    char c;
    int idx;
    int done = 0;

    if (argc > 1) {
        tapefile = argv[1];
    } else {
        msg("ERROR", "NO SPECIFIED TAPE FILE", -30);
        exit(-30);
    }

    HEADER();
    load_tapefile(tapefile);
    HEADER();

    print_action_table();

    for(;;) {
        done = halting_tests();
        if ( done > 0 ) {
            c = readchar();
            idx = nextstate(c);
            update_state(c, idx);
            print_tape_state();
        } else {
            print_tape_state();
            break;
        }
        HEADER();
    }
    printf( "%s", TAPE_STATE.output);

    exit(0);
}
/* ********************************************************************************************** */


